core/src/main/java/org/elasticsearch/cluster/routing/AllocationId.java

```diff
@@ -198,6 +201,6 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     }
 
     public static AllocationId fromXContent(XContentParser parser) throws IOException {
-        return ALLOCATION_ID_PARSER.parse(parser, new AllocationId.Builder()).build();
+        return ALLOCATION_ID_PARSER.parse(parser, new AllocationId.Builder(), () -> ParseFieldMatcher.STRICT).build();
```

Comment:
I tracked this one back to the place where the parser that gets passed down is created. The call-stack here is something like:
MetaDataStateFormat#loadLatestState -> ShardStateMetaData#fromXContent -> Allocation#fromXContent. 

I think we use ObjectParser a lot now in situations there when we initialize the parsing process we don't really have the settings available that we need to decide if we want to parse strict or without deprecation checks. This observation is indepentend of whether we want to store the ParseFieldMatcher in the parser or some sort of context. 
In this case I think we should default to STRICT because these json objects seem to be only created internally, so we should control that we never output anything deprecated.